File: frontend/src/types/student_project.ts
/**
 * Student Project Types
 * All field names use snake_case to match backend API
 */

export const ProjectStatus = {
  DRAFT: 'draft',
  IN_PROGRESS: 'in_progress',
  COMPLETED: 'completed',
  SUBMITTED: 'submitted',
} as const

export type ProjectStatus = typeof ProjectStatus[keyof typeof ProjectStatus]

export const ProjectStage = {
  ENGAGE: 'engage',
  EXPLORE: 'explore',
  EXPLAIN: 'explain',
  ELABORATE: 'elaborate',
  EVALUATE: 'evaluate',
} as const

export type ProjectStage = typeof ProjectStage[keyof typeof ProjectStage]

export interface ProjectCell {
  id?: number
  project_id?: number
  stage: ProjectStage
  cell_type: string
  title?: string
  content: any
  config?: any
  order: number
  created_at?: string
  updated_at?: string
}

export interface StudentProject {
  id: number
  title: string
  description?: string
  creator_id: number
  creator_name?: string
  project_type?: string
  status: ProjectStatus
  
  // 5E stage content
  engage_content: ProjectCell[]
  explore_content: ProjectCell[]
  explain_content: ProjectCell[]
  elaborate_content: ProjectCell[]
  evaluate_content: ProjectCell[]
  
  completion: {
    engage: number
    explore: number
    explain: number
    elaborate: number
    evaluate: number
  }
  
  cover_image_url?: string
  tags: string[]
  is_team_project: boolean
  team_members: number[]
  
  created_at: string
  updated_at: string
  submitted_at?: string
}

export interface StudentProjectCreate {
  title: string
  description?: string
  project_type?: string
}

export interface StudentProjectUpdate {
  title?: string
  description?: string
  status?: ProjectStatus
  cover_image_url?: string
  tags?: string[]
  engage_content?: any[]
  explore_content?: any[]
  explain_content?: any[]
  elaborate_content?: any[]
  evaluate_content?: any[]
}

export interface StudentProjectListParams {
  page?: number
  page_size?: number
  status?: ProjectStatus
}

export interface StudentProjectListResponse {
  items: StudentProject[]
  total: number
  page: number
  page_size: number
}

export interface StageContentUpdate {
  content: any[]
}

File: backend/app/schemas/student_project.py
"""
学生项目Schemas
"""

from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, field_validator

from app.models.student_project import ProjectStatus


class StudentProjectBase(BaseModel):
    """项目基础Schema"""

    title: str = Field(..., min_length=1, max_length=200, description="项目标题")
    description: Optional[str] = Field(None, description="项目描述")
    project_type: Optional[str] = Field(None, description="项目类型")


class StudentProjectCreate(StudentProjectBase):
    """项目创建Schema"""

    pass


class StudentProjectUpdate(BaseModel):
    """项目更新Schema"""

    title: Optional[str] = Field(None, min_length=1, max_length=200, description="项目标题")
    description: Optional[str] = Field(None, description="项目描述")
    status: Optional[ProjectStatus] = Field(None, description="项目状态")
    cover_image_url: Optional[str] = Field(None, max_length=500, description="封面图片URL")
    tags: Optional[List[str]] = Field(None, description="项目标签列表")
    # 5E阶段内容更新
    engage_content: Optional[List[Dict[str, Any]]] = Field(None, description="Engage阶段内容")
    explore_content: Optional[List[Dict[str, Any]]] = Field(None, description="Explore阶段内容")
    explain_content: Optional[List[Dict[str, Any]]] = Field(None, description="Explain阶段内容")
    elaborate_content: Optional[List[Dict[str, Any]]] = Field(None, description="Elaborate阶段内容")
    evaluate_content: Optional[List[Dict[str, Any]]] = Field(None, description="Evaluate阶段内容")

    @field_validator("engage_content", "explore_content", "explain_content", "elaborate_content", "evaluate_content", mode="before")
    @classmethod
    def validate_content(cls, v):
        """验证阶段内容字段"""
        if v is None:
            return None
        if not isinstance(v, list):
            return []
        return v


class StudentProjectResponse(StudentProjectBase):
    """项目响应Schema"""

    id: int
    creator_id: int
    creator_name: Optional[str] = None
    project_type: Optional[str] = None
    status: ProjectStatus
    completion: Dict[str, int] = Field(default_factory=dict, description="各阶段完成度")
    engage_content: List[Dict[str, Any]] = Field(default_factory=list, description="Engage阶段内容")
    explore_content: List[Dict[str, Any]] = Field(default_factory=list, description="Explore阶段内容")
    explain_content: List[Dict[str, Any]] = Field(default_factory=list, description="Explain阶段内容")
    elaborate_content: List[Dict[str, Any]] = Field(default_factory=list, description="Elaborate阶段内容")
    evaluate_content: List[Dict[str, Any]] = Field(default_factory=list, description="Evaluate阶段内容")
    cover_image_url: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    is_team_project: bool = False
    team_members: List[int] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    submitted_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class StudentProjectListResponse(BaseModel):
    """项目列表响应Schema"""

    items: List[StudentProjectResponse]
    total: int
    page: int
    page_size: int


class StageContentUpdate(BaseModel):
    """阶段内容更新Schema"""

    content: List[Dict[str, Any]] = Field(..., description="阶段内容列表")

    @field_validator("content", mode="before")
    @classmethod
    def validate_content(cls, v):
        """验证内容字段"""
        if v is None:
            return []
        if not isinstance(v, list):
            return []
        return v

